### PR TITLE
Banyandb storage plugin

### PR DESCRIPTION
**Pull request by lujiajing1126**
- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [ ] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/CHANGES.md).

### CODE REVIEW

**Comment by wu-sheng** [comment]
We have the image, please use infra e2e gha to add a new test.
You could take satellite project e2e as a reference.

**Comment by wu-sheng** [comment]
We only need to verify traces.

File: .github/workflows/e2e.storages.yaml [hunk 1]
```diff
@@ -36,7 +36,7 @@ jobs:
     timeout-minutes: 90
     strategy:
       matrix:
-        storage: ['mysql', 'es6', 'es7.0', 'es7.10', 'es7.14', 'opensearch', 'influxdb', 'tidb', 'postgresql']
+        storage: ['mysql', 'es6', 'es7.0', 'es7.10', 'es7.14', 'opensearch', 'influxdb', 'tidb', 'postgresql', 'banyandb']
```
Comment (wu-sheng) [inline .github/workflows/e2e.storages.yaml:39]: This e2e test would verify all metrics, service register. I think you only make traces working.

Comment (lujiajing1126) [reply .github/workflows/e2e.storages.yaml]: Sure

File: .github/workflows/e2e.banyandb.yaml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+          echo "::group::Copy Dist Package to test/e2e"
+          cp -R dist test/e2e/
+          echo "::endgroup::"
+      - name: Prepare E2E Test
+        shell: bash
+        run: |
+          echo "::group::Install SNAPSHOT apm-application-toolkit"
+          ./mvnw -DskipTests -N install
+          ./mvnw -f apm-application-toolkit -DskipTests -am install
+          echo "::endgroup::"
```
Comment (wu-sheng) [inline .github/workflows/e2e.banyandb.yaml:57]: You don't need this.

File: test/e2e/e2e-test/docker/storage/banyandb/expected/traces.yml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+traces:
+  - key: not null
+    endpointNames:
+      - /users
+    duration: ge 0
+    start: gt 0
+    isError: false
+    traceIds:
+      - not null
```
Comment (wu-sheng) [inline test/e2e/e2e-test/docker/storage/banyandb/expected/traces.yml:24]: You should check all fields of trace to make sure the Database is working as expected.

File: .github/workflows/e2e.banyandb.yaml (hunk 2 above)
Comment (lujiajing1126) [reply .github/workflows/e2e.banyandb.yaml:57]: I tried to run this locally. But I found that if I need to compile the e2e-provider, I have to install the `apm-application-toolkit` in the local repository.

File: test/e2e/e2e-test/docker/storage/banyandb/expected/traces.yml (hunk 3 above)
Comment (lujiajing1126) [reply test/e2e/e2e-test/docker/storage/banyandb/expected/traces.yml:24]: Sure. I will complete this later.

File: .github/workflows/e2e.banyandb.yaml (hunk 2 above)
Comment (wu-sheng) [reply .github/workflows/e2e.banyandb.yaml:57]: @kezhenxu94 Please give some suggestion. I think we need a guide line for new e2e-infra integration, which is already using in this branch.

**Comment by lujiajing1126** [comment]
It seems the workflow I've added did not take effect.

**Comment by wu-sheng** [comment]
> It seems the workflow I've added did not take effect.

I guess we need a separate pull request to inject a no logic GHA control file first, right? @kezhenxu94

**Comment by kezhenxu94** [comment]
> > It seems the workflow I've added did not take effect.
> 
> I guess we need a separate pull request to inject a no logic GHA control file first, right? @kezhenxu94

It did run, only because the file syntax is wrong.

Screen Shot 2021-08-20 at 08 32 50

File: oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/dao/BanyanDBTraceQueryDAO.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+            BasicTrace trace = new BasicTrace();
+            trace.setDuration(((Long) entity.getFields().get(3).getValue()).intValue());
+            trace.setStart(String.valueOf(entity.getFields().get(4).getValue()));
+            trace.setSegmentId(entity.getId());
+            trace.setError(((Long) entity.getFields().get(1).getValue()).intValue() == 1);
+            trace.getTraceIds().add((String) entity.getFields().get(0).getValue());
+            trace.getEndpointNames().add(IDManager.EndpointID.analysisId(
+                    (String) entity.getFields().get(2).getValue()
+            ).getEndpointName());
+            return trace;
```
Comment (wu-sheng) [inline oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/dao/BanyanDBTraceQueryDAO.java:127]: Please make sure the codes polished. When you face a 0-N field indices, you should make sure they are ordered.

Comment (wu-sheng) [reply oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/dao/BanyanDBTraceQueryDAO.java:127]: And where is the request fields? I am not sure how BanyanDB server implements the logic now when they are loss, but we access them through indices, they must be declared. If not, BanyanDB should response nothing.

**Comment by lujiajing1126** [comment]
@mrproliu 

I am using `${provider_host}` as a placeholder to get IP and port for the services defined in `docker-compose` file. But It seems that all placeholders are returning blank string. Also I've tested on both macOS and Linux.

Is this related to https://github.com/apache/skywalking-infra-e2e/pull/36?

**Comment by mrproliu** [comment]
> @mrproliu
> 
> image
> 
> I am using `${provider_host}` as a placeholder to get IP and port for the services defined in `docker-compose` file. But It seems that all placeholders are returning blank string. Also it can be reproducible on both macOS and Linux.

You need to declare which ports need to export to the environment, please add port declare into `docker-compose.yaml` file, here is a [demo link](https://github.com/apache/skywalking-satellite/blob/main/test/e2e/case/native-protocols/docker-compose.yml#L50-L51). 

> Is this related to [apache/skywalking-infra-e2e#36](https://github.com/apache/skywalking-infra-e2e/pull/36)?

After you change the `docker-compose.yml` port declare, the E2E may still fail, this PR it's to fix this problem, that's two different problems. After this PR has been merge, I will mention you to update the commit id. 

> Another feature request: can we provide an `env` field that allow to pass environments downwards into the container in Github Action environment.

Could you do more explain on this? which environment field do you want?

**Comment by lujiajing1126** [comment]
> > Another feature request: can we provide an `env` field that allow to pass environments downwards into the container in Github Action environment.
> 
> Could you do more explain on this? which environment field do you want?

In the current E2E workflow, `SW_AGENT_JDK_VERSION` is defined to indicate the JDK version we intend to use for the test.

https://github.com/apache/skywalking/blob/master/.github/workflows/e2e.storages.yaml#L24-L25

This variable is used in the `docker-compose.yml` file as a docker-build argument, i.e. `FROM adoptopenjdk/openjdk${SW_AGENT_JDK_VERSION}:alpine`.

https://github.com/apache/skywalking/blob/master/test/e2e/e2e-test/docker/base-compose.yml#L85-L86

and,

https://github.com/apache/skywalking/blob/master/test/e2e/e2e-test/docker/Dockerfile.provider#L16-L19

But this feature is missing in the state-of-art sw-infra-e2e framework according to the `action.yml`.

**Comment by mrproliu** [comment]
> > > Another feature request: can we provide an `env` field that allow to pass environments downwards into the container in Github Action environment.
> > 
> > 
> > Could you do more explain on this? which environment field do you want?
> 
> @mrproliu
> 
> In the current E2E workflow, `SW_AGENT_JDK_VERSION` is defined to indicate the JDK version we intend to use for the test.
> 
> https://github.com/apache/skywalking/blob/cf03304a48a940f5d599b3a2cbd0711b008c17d1/.github/workflows/e2e.storages.yaml#L24-L25
> 
> This variable is used in the `docker-compose.yml` file as a docker-build argument, i.e. `FROM adoptopenjdk/openjdk${SW_AGENT_JDK_VERSION}:alpine`.
> 
> https://github.com/apache/skywalking/blob/cf03304a48a940f5d599b3a2cbd0711b008c17d1/test/e2e/e2e-test/docker/base-compose.yml#L85-L86
> 
> and,
> 
> https://github.com/apache/skywalking/blob/cf03304a48a940f5d599b3a2cbd0711b008c17d1/test/e2e/e2e-test/docker/Dockerfile.provider#L16-L19
> 
> But this feature is missing in the state-of-art sw-infra-e2e framework according to the `action.yml`.
> 
> https://github.com/apache/skywalking-infra-e2e/blob/de5be41596f27867a15ab75dbad97ea538d863b8/action.yaml#L31-L35
> 
> Ping @kezhenxu94 . Do you have other consideration for this scenario? for example this is an intentional design

Thanks for your explain, Do you mean you want to correlate the GitHub Action environment into the Infra-E2E container? I think we missed that, so we can't correlate the GitHub Action Environment into E2E for now. I think we could dump the environment and use `--env-file` to propagate it.

**Comment by kezhenxu94** [comment]
> Ping @kezhenxu94 . Do you have other consideration for this scenario? for example this is an intentional design

I see. All env vars in GitHub actions should be (expected to be) passed downwards to e2e program, but as we use container in GHA the env vars are not implicitly passed so we need to do that manually. @mrproliu Seems more and more problems arise after we use Docker container in GHA.

**Comment by wu-sheng** [comment]
I am feeling we are overheating skywalking-infra-e2e. It is a young and unreleased project, we shouldn't expect its as powerful as others.
If this is not a block, let's create an issue for it to track, and set the env in the docker compose directly.

**Comment by mrproliu** [comment]
>  @mrproliu Seems more and more problems arise after we use Docker container in GHA.

I think it is only unstable now, but as we have more and more cases, we can make it more stable. After we released Infra-E2E and polish the docker image, it was worthwhile because we can speed up E2E testing. WDYT?

**Comment by lujiajing1126** [comment]
> I am feeling we are overheating skywalking-infra-e2e. It is a young and unreleased project, we shouldn't expect its as powerful as others.
> If this is not a block, let's create an issue for it to track, and set the env in the docker compose directly.

Env is not a blocking issue. I've made a workaround.

**Comment by wu-sheng** [comment]
You could submit an issue directly with context, then you or someone else could enhance that young project later.

**Comment by lujiajing1126** [comment]
Another issue report:

```
$ swctl --display yaml --base-url=http://127.0.0.1:49184/graphql trace ls
FATAL graphql: The variables input contains a field name 'endpointName' that is not defined for input object type 'TraceQueryCondition'
```

@wu-sheng is this related to the recent removal of `endpointName` in https://github.com/apache/skywalking/pull/7438

```
$ swctl -v
swctl version 0.7.0
```

**Comment by wu-sheng** [comment]
> Another issue report:
> 
> ```
> $ swctl --display yaml --base-url=http://127.0.0.1:49184/graphql trace ls
> FATAL graphql: The variables input contains a field name 'endpointName' that is not defined for input object type 'TraceQueryCondition'
> ```
> 
> @wu-sheng is this related to the recent removal of `endpointName` in #7438
> 
> ```
> $ swctl -v
> swctl version 0.7.0
> ```

@kezhenxu94 Could you check whether we have removed this from CLI? If not, please do.

File: oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/dao/BanyanDBTraceQueryDAO.java (hunk 4 above)
Comment (lujiajing1126) [reply oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/dao/BanyanDBTraceQueryDAO.java]: > Please make sure the codes polished. When you face a 0-N field indices, you should make sure they are ordered.

Sorted.

Comment (lujiajing1126) [reply oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/dao/BanyanDBTraceQueryDAO.java]: > And where is the request fields? I am not sure how BanyanDB server implements the logic now when they are loss, but we access them through indices, they must be declared. If not, BanyanDB should response nothing.

What do you mean by request fields here?

File: test/e2e/e2e-test/docker/storage/banyandb/expected/traces.yml (hunk 3 above)
Comment (lujiajing1126) [reply test/e2e/e2e-test/docker/storage/banyandb/expected/traces.yml]: More test cases are added

**Comment by lujiajing1126** [comment]
BanyanDB E2E has passed. Pls have a review again @wu-sheng @hanahmily @kezhenxu94

File: oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/BanyanDBStorageProvider.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        this.registerServiceImplementation(
+                IProfileThreadSnapshotQueryDAO.class, new BanyanDBProfileThreadSnapshotQueryDAO());
+        this.registerServiceImplementation(UITemplateManagementDAO.class, new BanyanDBUITemplateManagementDAO());
+
+        this.registerServiceImplementation(IEventQueryDAO.class, new BanyanDBEventQueryDAO());
+    }
+
+    @Override
+    public void start() throws ServiceNotProvidedException, ModuleStartException {
+        final ConfigService configService = getManager().find(CoreModule.NAME)
```
Comment (kezhenxu94) [inline oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/BanyanDBStorageProvider.java:127]: Unused

**Review by kezhenxu94 — COMMENTED** [review]
Just some nits

**Comment by wu-sheng** [comment]
@kezhenxu94 Does GitHub provide a snapshot based maven repo? If so, we could keep what we did for agent test, which is, We could create a `banyandb-java-client` repo, and hosts all Java driver related codes, and get a dev maven repo too. In the main repo, we are going to need snapshot dependency only.

After that, we could keep this PR to the master branch(don't need to sync codes anymore), every time this PR gets update, the test could run directly.

@lujiajing1126 @hanahmily What do you think about this, rather than maintaining another branch?

**Comment by kezhenxu94** [comment]
> @kezhenxu94 Does GitHub provide a snapshot based maven repo?

We can do that

**Comment by hanahmily** [comment]
It makes sense to me. Maybe the Banyandb repo is a proper place to hold
client Java codes and build artifacts.

**Comment by lujiajing1126** [comment]
> @kezhenxu94 Does GitHub provide a snapshot based maven repo? If so, we could keep what we did for agent test, which is, We could create a `banyandb-java-client` repo, and hosts all Java driver related codes, and get a dev maven repo too. In the main repo, we are going to need snapshot dependency only.
> 
> After that, we could keep this PR to the master branch(don't need to sync codes anymore), every time this PR gets update, the test could run directly.
> 
> @lujiajing1126 @hanahmily What do you think about this, rather than maintaining another branch?

Make sense to me.

**Comment by kezhenxu94** [comment]
This might be a little late as I didn't review the first few pull requests to this branch `banyandb-integration`, later that I reviewed the merged codes and found that the so-called `banyandb-java-client` is just a simple wrapper of gRPC client, which makes me feel whether we really need to maintain this kind of `java-client`:

- protobuffers itself already serves as a protocol layer / API layer, open source projects usually define protobuffers schema and let users compile and use gRPC to call the services, instead of building another wrapper on top of gRPC.
- when we wrap gRPC in our java-client, it means users have less possibility to customize the gRPC configurations, or, we have to expose gRPC configurations in "our own SDK", which is not a good idea in terms of OOP capsulation. Also, dependency management is what we should consider.
- currently, the `banyandb-java-client` depends on our `apm-datacarrier` and queue the requests ourselves, I think the gRPC itself must have queue internally and know when is the best time to flush the requests.

But, I don't mean the module `banyandb-java-client` is useless, it's useful in oap-server, but I'm kind of concerned whether it's useful for the BanyanDB end users and so is it necessary to maintain and publish it separately .

**Comment by wu-sheng** [comment]
> currently, the banyandb-java-client depends on our apm-datacarrier and queue the requests ourselves, I think the gRPC itself must have queue internally and know when is the best time to flush the requests.

Let me get to this. gRPC and netty have internal queue, but it is for direct memory for sending, which can't avoid OOM :) We have suffered those years ago.

**Comment by wu-sheng** [comment]
> but I'm kind of concerned whether it's useful for the BanyanDB end users and so is it necessary to maintain and publish it separately .

We don't know, but when you dig deeper, you will notice. BanyanDB Java client is not only for providing client implementation, it is scoping the client proto files. All those proto files in the server side are not all cared of by the client. We show how to integrate. If end users will(long future will) run some integration, they could use the proto files in client folder, and go further.

**Comment by wu-sheng** [comment]
Every client is a wrapper for a protocol, like you are doing our new ElasticSearch-protocol client, or what Elastic team did and we are using right now. This is not a condition we should or should not have a client repository. :)

**Comment by wu-sheng** [comment]
> Every client is a wrapper for a protocol, like you are doing our new ElasticSearch-protocol client, or what Elastic team did and we are using right now. This is not a condition we should or should not have a client repository. :)

The value of client repository is making people easier to understand and use the network protocol. Think in this way, our trace protocol is out there for years. But you and only several agents' founders have/had read/understood the protocol field by field. But thousands users out there, they care about language-specific APIs only.

**Comment by wu-sheng** [comment]
> protobuffers itself already serves as a protocol layer / API layer, open source projects usually define protobuffers schema and let users compile and use gRPC to call the services, instead of building another wrapper on top of gRPC.

That is and only for professional cross open source projects integration. Like we analysis Envoy protobuf, and have our complex solution for mesh observability. But that is a very rare case and for a small subset of all open source users.
We need those proto, and we need more.

**Comment by kezhenxu94** [comment]
> We don't know, but when you dig deeper, you will notice. BanyanDB Java client is not only for providing client implementation, it is scoping the client proto files. All those proto files in the server side are not all cared of by the client. We show how to integrate. If end users will(long future will) run some integration, they could use the proto files in client folder, and go further.

IMHO that's our responsibility to make it clear in repo apache/skywalking-banyandb or apache/skywalking-banyandb-api (instead of in banyandb-java-client repo) what are client side `.proto` files and what are server side `.proto` files, even we have Java client, it should not take the responsibility to scope the client `.proto` files, after all, users may use other programming languages.

> That is and only for professional cross open source projects integration. Like we analysis Envoy protobuf, and have our complex solution for mesh observability. But that is a very rare case and for a small subset of all open source users.
We need those proto, and we need more.

That's my point, BanyanDB is an APM-specific database, not general-purpose database, I just thought that there won't be many ordinary users integrate it.

I'm totally OK to have `banyandb-java-client` repo, just wanted to raise the discussion above as soon as I notice it.

**Comment by lujiajing1126** [comment]
> > currently, the banyandb-java-client depends on our apm-datacarrier and queue the requests ourselves, I think the gRPC itself must have queue internally and know when is the best time to flush the requests.
> 
> Let me get to this. gRPC and netty have internal queue, but it is for direct memory for sending, which can't avoid OOM :) We have suffered those years ago.

BTW, I've recently found a project, https://github.com/salesforce/reactive-grpc. 

As I understand, according to the document, this project can properly handle the so-called back-pressure issue which is similar to what we provide thru `apm-data-carrier`. And it seems to be more capable since it is deeply integrated to the transport layer, i.e. HTTP/2 flow control.

**Comment by wu-sheng** [comment]
> IMHO that's our responsibility to make it clear in repo apache/skywalking-banyandb or apache/skywalking-banyandb-api (instead of in banyandb-java-client repo) what are client side .proto files and what are server side .proto files, even we have Java client, it should not take the responsibility to scope the client .proto files, after all, users may use other programming languages.

We should not. But we are for short term. We have not gotten the clear boundary between client/network proto and server internal proto. They are sharing. We will have `apache/skywalking-banyandb-api ` eventually, but not now.

> That's my point, BanyanDB is an APM-specific database, not general-purpose database, I just thought that there won't be many ordinary users integrate it.

I think it wouldn't. But who knows, like many things we did, we didn't expect they are being used widely. But you found yesterday e2e partial code copying case, who knows. SkyWalking is being watched by countless projects' maintainers and users.

**Comment by wu-sheng** [comment]
> BTW, I've recently found a project, https://github.com/salesforce/reactive-grpc.
> 
> As I understand, according to the document, this project can properly handle the so-called back-pressure issue which is similar to what we provide thru `apm-data-carrier`. And it seems to be more capable since it is deeply integrated to the transport layer, i.e. HTTP/2 flow control.

We could try if you want. But you have to resolve the dependency tree, and notice, that is also a vendor-base OSS with very limited activities(https://github.com/salesforce/reactive-grpc/pulse/monthly). Saleforce doesn't have very famous and popular OSS libs, we may face out-of-maintain case.

**Comment by wu-sheng** [comment]
Repo created https://github.com/apache/skywalking-banyandb-java-client

**Comment by wu-sheng** [comment]
@lujiajing1126 I am going to move the client codebase in your fork into that repo. Is that OK? We are still waiting for banyandb server side change right?

**Comment by lujiajing1126** [comment]
> @lujiajing1126 I am going to move the client codebase in your fork into that repo. Is that OK? We are still waiting for banyandb server side change right?

I think it's OK to move now. Yes, we are still waiting for the new index module.

**Comment by wu-sheng** [comment]
@lujiajing1126 All source codes are there. Once @kezhenxu94 have time to set up snapshot maven deploy in the CI process, we could use it directly. 

This PR should be closed. When everything(BanyanDB server) is ready, we should start one targeting to master branch.

**Comment by kezhenxu94** [comment]
We can also deploy those snapshots to maven central via github workflow, which is allowed in infra , if I remember correctly

**Comment by wu-sheng** [comment]
> If that's unacceptable, we can also deploy those snapshots to maven central via github workflow, which is allowed in infra , if I remember correctly

You could do deploy on https://repository.apache.org/, but not Maven central.

**Comment by kezhenxu94** [comment]
> > If that's unacceptable, we can also deploy those snapshots to maven central via github workflow, which is allowed in infra , if I remember correctly
> 
> You could do deploy on https://repository.apache.org/, but not Maven central.

Ticket is created to request infra team to set up the needed token https://issues.apache.org/jira/browse/INFRA-22286

**Comment by kezhenxu94** [comment]
@lujiajing1126 snapshot is deployed, you need to add the following snippet to main repo (`oap-server-bom` / `oap-server-bom-es7`) when you want to reference it

```xml
    <repositories>
        <repository>
            <id>apache-snapshot</id>
            <name>Apache Snapshot</name>
            <url>https://repository.apache.org/content/repositories/snapshots</url>
        </repository>
    </repositories>
```

```xml
        <dependency>
            <groupId>org.apache.skywalking</groupId>
            <artifactId>banyandb-java-client</artifactId>
            <version>0.1.0-SNAPSHOT</version>
        </dependency>
```

**Comment by lujiajing1126** [comment]
> @lujiajing1126 snapshot is deployed, you need to add the following snippet to main repo (`oap-server-bom` / `oap-server-bom-es7`) when you want to reference it
> 
> ```
> <project>
> ...
>     <repositories>
>         <repository>
>             <id>apache-snapshot</id>
>             <name>Apache Snapshot</name>
>             <url>https://repository.apache.org/content/repositories/snapshots</url>
>         </repository>
>     </repositories>
> ```
> 
> and use it in the `pom.xml` where you want to use it
> 
> ```
>         <dependency>
>             <groupId>org.apache.skywalking</groupId>
>             <artifactId>banyandb-java-client</artifactId>
>             <version>0.1.0-SNAPSHOT</version>
>         </dependency>
> ```

Got it. Thanks

**Pull request closed without merging**